{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "193a5c5d",
   "metadata": {},
   "source": [
    "# Run *de novo*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6f680e1",
   "metadata": {},
   "source": [
    "### alt80"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ffd1050",
   "metadata": {},
   "outputs": [],
   "source": [
    "denovo_map.pl --samples /data/users/sousal/03_compare_methods/02_filtered-adaptor-rmv-q30-l60-L60  --popmap ../popmap-alt80 -o ./alti -M 3 -n 3 -T 14\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "37e5c8cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logging to '04_denovo/alti_populations/populations.log'.\n",
      "Locus/sample distributions will be written to '04_denovo/alti_populations/populations.log.distribs'.\n",
      "populations parameters selected:\n",
      "  Percent samples limit per population: 0\n",
      "  Locus Population limit: 1\n",
      "  Percent samples overall: 0.8\n",
      "  Minor allele frequency cutoff: 0.05\n",
      "  Maximum observed heterozygosity cutoff: 1\n",
      "  Applying Fst correction: none.\n",
      "  Pi/Fis kernel smoothing: off\n",
      "  F-stats kernel smoothing: off\n",
      "  Bootstrap resampling: off\n",
      "\n",
      "Parsing population map...\n",
      "The population map contained 80 samples, 3 population(s), 1 group(s).\n",
      "Working on 80 samples.\n",
      "Working on 3 population(s):\n",
      "    north: 2062150, 2062151, 2062152, 2062154, 2062164, 2062165, 2062166, 2062167, 2062168, 2062169, 2062170, 2062171, 2062172, 2062173, \n",
      "           2062174, 2062175, 2062176, 2062177, 2062178, 2062179, 2062180, 2062181, 2062182, 2062183, 2062185, 2062186, 2062187, 2062343\n",
      "    south: 2062188, 2062189, 2062191, 2062192, 2062196, 2062198, 2062199, 2062200, 2062202, 2062204, 2062205, 2062206, 2062207, 2062208, \n",
      "           2062209, 2062210, 2062211, 2062212, 2062213, 2062215, 2062216, 2062217, 2062218, 2062219, 2062220, 2062221, 2062222, 2062223, \n",
      "           2062224, 2062225, 2062226, 2062230, 2062231, 2062232, 2062233, 2062234, 2062235, 2062236, 2062237, 2062238, 2062239, 2070207\n",
      "    west: 2062155, 2062156, 2062157, 2062158, 2062227, 2062228, 2062229, 2062240, 2062241, 2062242\n",
      "Working on 1 group(s) of populations:\n",
      "    defaultgrp: north, south, west\n",
      "\n",
      "SNPs and calls will be written in VCF format to '04_denovo/alti_populations/populations.snps.vcf'\n",
      "Raw haplotypes will be written to '04_denovo/alti_populations/populations.haplotypes.tsv'\n",
      "Population-level summary statistics will be written to '04_denovo/alti_populations/populations.sumstats.tsv'\n",
      "Population-level haplotype summary statistics will be written to '04_denovo/alti_populations/populations.hapstats.tsv'\n",
      "\n",
      "Processing data in batches:\n",
      "  * load a batch of catalog loci and apply filters\n",
      "  * compute SNP- and haplotype-wise per-population statistics\n",
      "  * compute F-statistics\n",
      "  * write the above statistics in the output files\n",
      "  * export the genotypes/haplotypes in specified format(s)\n",
      "More details in '04_denovo/alti_populations/populations.log.distribs'.\n",
      "\n",
      "Now processing...\n",
      "Batch 1 \n",
      "Batch 2 \n",
      "Batch 3 \n",
      "\n",
      "Removed 700481 loci that did not pass sample/population constraints from 721544 loci.\n",
      "Kept 21063 loci, composed of 1308254 sites; 72598 of those sites were filtered, 10323 variant sites remained.\n",
      "Mean genotyped sites per locus: 62.11bp (stderr 0.01).\n",
      "\n",
      "Population summary statistics (more detail in populations.sumstats_summary.tsv):\n",
      "  north: 25.273 samples per locus; pi: 0.25577; all/variant/polymorphic sites: 1308254/10323/9326; private alleles: 259\n",
      "  south: 35.932 samples per locus; pi: 0.24423; all/variant/polymorphic sites: 1308254/10323/9549; private alleles: 314\n",
      "  west: 8.9845 samples per locus; pi: 0.22454; all/variant/polymorphic sites: 1308130/10321/6670; private alleles: 277\n",
      "\n",
      "Population pair divergence statistics (more in populations.fst_summary.tsv and populations.phistats_summary.tsv):\n",
      "  north-south: mean Fst: 0.033759; mean Phi_st: 0.047697; mean Fst': 0.021551; mean Dxy: 0.0047187\n",
      "  north-west: mean Fst: 0.090284; mean Phi_st: 0.13643; mean Fst': 0.089867; mean Dxy: 0.0053721\n",
      "  south-west: mean Fst: 0.081877; mean Phi_st: 0.13881; mean Fst': 0.089703; mean Dxy: 0.0052677\n",
      "\n",
      "Populations is done.\n"
     ]
    }
   ],
   "source": [
    "!populations -P 04_denovo/alti -O 04_denovo/alti_populations -M popmap-alt80 -R 0.8 --min-maf 0.05 --write-single-snp --vcf --fstats -t 16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b764d78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: populations: command not found\r\n"
     ]
    }
   ],
   "source": [
    "!populations -P alti -O alti_populations_90 -M popmap-alt80 -R 0.9 --min-maf 0.05 --write-single-snp --vcf --fstats -t 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b05043a",
   "metadata": {},
   "source": [
    "### grp41"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edec6b61",
   "metadata": {},
   "outputs": [],
   "source": [
    "denovo_map.pl --samples /data/users/sousal/03_compare_methods/02_filtered-adaptor-rmv-q30-l100  --popmap ../popmap-grp41  -o ./grp -M 3 -n 3 -T 14\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ccce54c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logging to '04_denovo/grp_populations/populations.log'.\n",
      "Locus/sample distributions will be written to '04_denovo/grp_populations/populations.log.distribs'.\n",
      "populations parameters selected:\n",
      "  Percent samples limit per population: 0\n",
      "  Locus Population limit: 1\n",
      "  Percent samples overall: 0.8\n",
      "  Minor allele frequency cutoff: 0.05\n",
      "  Maximum observed heterozygosity cutoff: 1\n",
      "  Applying Fst correction: none.\n",
      "  Pi/Fis kernel smoothing: off\n",
      "  F-stats kernel smoothing: off\n",
      "  Bootstrap resampling: off\n",
      "\n",
      "Parsing population map...\n",
      "The population map contained 41 samples, 4 population(s), 1 group(s).\n",
      "Working on 41 samples.\n",
      "Working on 4 population(s):\n",
      "    Preit: 3005842, 3005852, 3005888, 3005891, 3005899, 3005900, 3005911, 3005912, 3005924, 3006912, 3006944, 3006949\n",
      "    Psaxi: 3005841, 3005853, 3005865, 3005877, 3005889, 3005901, 3005903\n",
      "    Psche: 2990934, 2990940, 3005839, 3005851, 3005863, 3005874, 3005875, 3005886, 3005887, 3006939, 3006940, 3006945\n",
      "    Pteri: 3005843, 3005854, 3005855, 3005867, 3005890, 3005902, 3005914, 3005926, 3006928, 3006934\n",
      "Working on 1 group(s) of populations:\n",
      "    defaultgrp: Preit, Psaxi, Psche, Pteri\n",
      "\n",
      "SNPs and calls will be written in VCF format to '04_denovo/grp_populations/populations.snps.vcf'\n",
      "Raw haplotypes will be written to '04_denovo/grp_populations/populations.haplotypes.tsv'\n",
      "Population-level summary statistics will be written to '04_denovo/grp_populations/populations.sumstats.tsv'\n",
      "Population-level haplotype summary statistics will be written to '04_denovo/grp_populations/populations.hapstats.tsv'\n",
      "\n",
      "Processing data in batches:\n",
      "  * load a batch of catalog loci and apply filters\n",
      "  * compute SNP- and haplotype-wise per-population statistics\n",
      "  * compute F-statistics\n",
      "  * write the above statistics in the output files\n",
      "  * export the genotypes/haplotypes in specified format(s)\n",
      "More details in '04_denovo/grp_populations/populations.log.distribs'.\n",
      "\n",
      "Now processing...\n",
      "Batch 1 \n",
      "\n",
      "Removed 116803 loci that did not pass sample/population constraints from 117224 loci.\n",
      "Kept 421 loci, composed of 42881 sites; 1113 of those sites were filtered, 257 variant sites remained.\n",
      "Mean genotyped sites per locus: 101.86bp (stderr 0.08).\n",
      "\n",
      "Population summary statistics (more detail in populations.sumstats_summary.tsv):\n",
      "  Preit: 10.117 samples per locus; pi: 0.11396; all/variant/polymorphic sites: 42881/257/93; private alleles: 14\n",
      "  Psaxi: 5.8444 samples per locus; pi: 0.11432; all/variant/polymorphic sites: 42881/257/70; private alleles: 8\n",
      "  Psche: 10.401 samples per locus; pi: 0.20496; all/variant/polymorphic sites: 42881/257/133; private alleles: 59\n",
      "  Pteri: 8.3074 samples per locus; pi: 0.19439; all/variant/polymorphic sites: 42881/257/124; private alleles: 55\n",
      "\n",
      "Population pair divergence statistics (more in populations.fst_summary.tsv and populations.phistats_summary.tsv):\n",
      "  Preit-Psaxi: mean Fst: 0.13736; mean Phi_st: 0.16665; mean Fst': 0.11308; mean Dxy: 0.0033931\n",
      "  Preit-Psche: mean Fst: 0.25758; mean Phi_st: 0.32925; mean Fst': 0.23727; mean Dxy: 0.004088\n",
      "  Preit-Pteri: mean Fst: 0.30348; mean Phi_st: 0.37493; mean Fst': 0.29039; mean Dxy: 0.0045898\n",
      "  Psaxi-Psche: mean Fst: 0.271; mean Phi_st: 0.34737; mean Fst': 0.26951; mean Dxy: 0.0045182\n",
      "  Psaxi-Pteri: mean Fst: 0.3111; mean Phi_st: 0.37737; mean Fst': 0.30866; mean Dxy: 0.0048538\n",
      "  Psche-Pteri: mean Fst: 0.22731; mean Phi_st: 0.30027; mean Fst': 0.20212; mean Dxy: 0.00395\n",
      "\n",
      "Populations is done.\n"
     ]
    }
   ],
   "source": [
    "!populations -P 04_denovo/grp -O 04_denovo/grp_populations -M popmap-grp41 -R 0.8 --min-maf 0.05 --write-single-snp --vcf --fstats -t 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9189060",
   "metadata": {},
   "source": [
    "### phylo29"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94099e0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "denovo_map.pl --samples /data/users/sousal/03_compare_methods/02_filtered-adaptor-rmv-q30-l60-L60  --popmap ../popmap-phylo29.txt  -o ./phylo -M 3 -n 3 -T 14 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8c4e07ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logging to '04_denovo/phylo_populations/populations.log'.\n",
      "Locus/sample distributions will be written to '04_denovo/phylo_populations/populations.log.distribs'.\n",
      "populations parameters selected:\n",
      "  Percent samples limit per population: 0\n",
      "  Locus Population limit: 1\n",
      "  Percent samples overall: 0.8\n",
      "  Minor allele frequency cutoff: 0.05\n",
      "  Maximum observed heterozygosity cutoff: 1\n",
      "  Applying Fst correction: none.\n",
      "  Pi/Fis kernel smoothing: off\n",
      "  F-stats kernel smoothing: off\n",
      "  Bootstrap resampling: off\n",
      "\n",
      "Parsing population map...\n",
      "The population map contained 29 samples, 5 population(s), 1 group(s).\n",
      "Working on 29 samples.\n",
      "Working on 5 population(s):\n",
      "    caes: caes1, caes12, caes3, caes7, caes8, caes9\n",
      "    duse: duse11, duse12, duse2, duse23, duse31, duse7\n",
      "    miss: miss10, miss36, miss38, miss39, miss40, miss58\n",
      "    para: para156, para16, para178, para227, para240, para80\n",
      "    send: send114, send117, send121, send54, send9\n",
      "Working on 1 group(s) of populations:\n",
      "    defaultgrp: caes, duse, miss, para, send\n",
      "\n",
      "SNPs and calls will be written in VCF format to '04_denovo/phylo_populations/populations.snps.vcf'\n",
      "Raw haplotypes will be written to '04_denovo/phylo_populations/populations.haplotypes.tsv'\n",
      "Population-level summary statistics will be written to '04_denovo/phylo_populations/populations.sumstats.tsv'\n",
      "Population-level haplotype summary statistics will be written to '04_denovo/phylo_populations/populations.hapstats.tsv'\n",
      "\n",
      "Processing data in batches:\n",
      "  * load a batch of catalog loci and apply filters\n",
      "  * compute SNP- and haplotype-wise per-population statistics\n",
      "  * compute F-statistics\n",
      "  * write the above statistics in the output files\n",
      "  * export the genotypes/haplotypes in specified format(s)\n",
      "More details in '04_denovo/phylo_populations/populations.log.distribs'.\n",
      "\n",
      "Now processing...\n",
      "Batch 1 \n",
      "Batch 2 \n",
      "\n",
      "Removed 228255 loci that did not pass sample/population constraints from 239731 loci.\n",
      "Kept 11476 loci, composed of 721463 sites; 25231 of those sites were filtered, 7691 variant sites remained.\n",
      "Mean genotyped sites per locus: 62.87bp (stderr 0.01).\n",
      "\n",
      "Population summary statistics (more detail in populations.sumstats_summary.tsv):\n",
      "  caes: 5.769 samples per locus; pi: 0.12035; all/variant/polymorphic sites: 721463/7691/2315; private alleles: 1097\n",
      "  duse: 4.9988 samples per locus; pi: 0.13869; all/variant/polymorphic sites: 721463/7691/2601; private alleles: 728\n",
      "  miss: 5.179 samples per locus; pi: 0.1413; all/variant/polymorphic sites: 721463/7691/2622; private alleles: 1067\n",
      "  para: 5.7162 samples per locus; pi: 0.14411; all/variant/polymorphic sites: 721463/7691/2915; private alleles: 1000\n",
      "  send: 4.5712 samples per locus; pi: 0.13583; all/variant/polymorphic sites: 721336/7689/2511; private alleles: 533\n",
      "\n",
      "Population pair divergence statistics (more in populations.fst_summary.tsv and populations.phistats_summary.tsv):\n",
      "  caes-duse: mean Fst: 0.31152; mean Phi_st: 0.3563; mean Fst': 0.28617; mean Dxy: 0.0075539\n",
      "  caes-miss: mean Fst: 0.35892; mean Phi_st: 0.41237; mean Fst': 0.34056; mean Dxy: 0.0081682\n",
      "  caes-para: mean Fst: 0.3323; mean Phi_st: 0.38378; mean Fst': 0.31924; mean Dxy: 0.0079122\n",
      "  caes-send: mean Fst: 0.31325; mean Phi_st: 0.35763; mean Fst': 0.29021; mean Dxy: 0.0076468\n",
      "  duse-miss: mean Fst: 0.27779; mean Phi_st: 0.31423; mean Fst': 0.2458; mean Dxy: 0.0072328\n",
      "  duse-para: mean Fst: 0.25483; mean Phi_st: 0.29037; mean Fst': 0.22907; mean Dxy: 0.0069901\n",
      "  duse-send: mean Fst: 0.22625; mean Phi_st: 0.2422; mean Fst': 0.18478; mean Dxy: 0.0067111\n",
      "  miss-para: mean Fst: 0.3121; mean Phi_st: 0.35976; mean Fst': 0.29449; mean Dxy: 0.0076697\n",
      "  miss-send: mean Fst: 0.26733; mean Phi_st: 0.29621; mean Fst': 0.23578; mean Dxy: 0.0072807\n",
      "  para-send: mean Fst: 0.26385; mean Phi_st: 0.29921; mean Fst': 0.24162; mean Dxy: 0.007184\n",
      "\n",
      "Populations is done.\n"
     ]
    }
   ],
   "source": [
    "!populations -P 04_denovo/phylo -O 04_denovo/phylo_populations -M popmap-phylo29 -R 0.8 --min-maf 0.05 --write-single-snp --vcf --fstats -t 16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95e0ce65",
   "metadata": {},
   "outputs": [],
   "source": [
    "/data/users/sousal/programs/stacks-2.64/tsv2bam -P $out_path $file_paths \n",
    "/data/users/sousal/programs/stacks-2.64/tsv2bam -P ./phylo -M ../popmap-phylo29 -p 14\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7295bc84",
   "metadata": {},
   "outputs": [],
   "source": [
    "for n in 0.5 0.9 0.95;\n",
    "    do\n",
    "    populations -P 04_denovo/phylo -O 04_denovo/phylo_populations -M popmap-phylo29 -R $n --min-maf 0.05 --write-single-snp --vcf --fstats -t 16\n",
    "    done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16ab5135",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "# Script mais simples baseado na sua estrutura exata\n",
    "\n",
    "# Grupos e popmaps\n",
    "groups=(\"alti\" \"grp\" \"phylo\")\n",
    "popmaps=(\"popmap-alti80\" \"popmap-grp41\" \"popmap-phylo29\")\n",
    "\n",
    "# Valores R e sufixos\n",
    "r_values=(\"0.5\" \"0.9\" \"0.95\")\n",
    "r_suffixes=(\"50\" \"90\" \"95\")\n",
    "\n",
    "# Executar em paralelo\n",
    "for i in \"${!groups[@]}\"; do\n",
    "    group=\"${groups[i]}\"\n",
    "    popmap=\"${popmaps[i]}\"\n",
    "    \n",
    "    for j in \"${!r_values[@]}\"; do\n",
    "        r_value=\"${r_values[j]}\"\n",
    "        r_suffix=\"${r_suffixes[j]}\"\n",
    "        \n",
    "        output_dir=\"${group}_populations_${r_suffix}\"\n",
    "        \n",
    "        echo \"ðŸš€ Executando: $group -> $output_dir (R=$r_value)\"\n",
    "        \n",
    "        populations -P \"$group\"/ -O \"$output_dir\" -M \"$popmap\" -R \"$r_value\" \\\n",
    "            --min-maf 0.05 --write-single-snp --vcf --fstats -t 16 &\n",
    "    done\n",
    "done\n",
    "\n",
    "# Esperar todos os processos\n",
    "wait\n",
    "echo \"âœ… Todos os processos completados!\""
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
